File: src/lib/androidFileBridge.ts
import { Capacitor } from "@capacitor/core";

// Android-only bridge to MainActivity's AndroidFileBridge (window.ElfAndroidFiles).
// The Android WebView has no download manager, no Web Share API and no PDF
// renderer, so exports/attachments are handed to native code, which writes
// them to the app cache and launches the system share sheet / a viewer app.
// Everything here is inert on iOS and in browsers (isAndroidNativeApp() is
// false), which keep their existing paths.

export type AndroidBridgeErrorCode =
  | "bridge_unavailable"
  | "no_handler"
  | "forbidden_origin"
  | "empty_data"
  | "too_large"
  | "bad_mime"
  | "bad_data"
  | "io_error"
  | "launch_failed";

export class AndroidFileBridgeError extends Error {
  code: AndroidBridgeErrorCode | string;
  constructor(code: string, message: string) {
    super(message);
    this.name = "AndroidFileBridgeError";
    this.code = code;
  }
}

type BridgeResult = { ok: true; fileName?: string } | { ok: false; error: string };

type ElfAndroidFiles = {
  saveAndShare(base64: string, fileName: string, mimeType: string): string;
  openWithViewer(base64: string, fileName: string, mimeType: string): string;
};

declare global {
  interface Window {
    ElfAndroidFiles?: ElfAndroidFiles;
  }
}

const MAX_BYTES = 25 * 1024 * 1024;
const ALLOWED_MIME = new Set([
  "application/pdf",
  "text/csv",
  "text/calendar",
  "image/png",
  "image/jpeg",
  "image/webp",
  "image/gif",
  "application/octet-stream",
]);
const EXTENSION_FOR_MIME: Record<string, string> = {
  "application/pdf": "pdf",
  "text/csv": "csv",
  "text/calendar": "ics",
  "image/png": "png",
  "image/jpeg": "jpg",
  "image/webp": "webp",
  "image/gif": "gif",
};

const USER_MESSAGES: Record<string, string> = {
  bridge_unavailable: "File sharing isn't available in this version of the app. Please update the app.",
  no_handler: "No app is installed that can open this file. Try saving or sharing it instead.",
  forbidden_origin: "This page isn't allowed to share files.",
  empty_data: "The file is empty.",
  too_large: "This file is too large to share from the app.",
  bad_mime: "This file type isn't supported.",
  bad_data: "The file couldn't be read.",
  io_error: "The file couldn't be prepared. Please try again.",
  launch_failed: "Couldn't open the share sheet. Please try again.",
};

/** True only inside the installed Android app (never iOS, never a browser). */
export function isAndroidNativeApp(): boolean {
  return Capacitor.isNativePlatform() && Capacitor.getPlatform() === "android";
}

/** Mirrors AndroidFileBridge.sanitizeFileName so callers can predict the name. */
export function sanitizeFileName(raw: string | null | undefined, mimeType?: string): string {
  let name = (raw ?? "").toString();
  const slash = Math.max(name.lastIndexOf("/"), name.lastIndexOf("\\"));
  if (slash >= 0) name = name.slice(slash + 1);
  name = name.replace(/[^A-Za-z0-9._ -]/g, "_");
  name = name.replace(/\.{2,}/g, ".");
  name = name.replace(/^[. ]+/, "").trim();
  if (name.length > 100) {
    const dot = name.lastIndexOf(".");
    const ext = dot > 0 && name.length - dot <= 10 ? name.slice(dot) : "";
    name = name.slice(0, 100 - ext.length) + ext;
  }
  if (!name) name = "file";
  if (!name.includes(".")) {
    const ext = mimeType ? EXTENSION_FOR_MIME[normalizeMimeType(mimeType) ?? ""] : undefined;
    if (ext) name = `${name}.${ext}`;
  }
  return name;
}

/**
 * Lower-cases, strips parameters ("; charset=utf-8") and validates the type.
 * Returns null for malformed input; well-formed types outside the allowlist
 * collapse to application/octet-stream (never trust an arbitrary type).
 */
export function normalizeMimeType(raw: string | null | undefined): string | null {
  const base = (raw ?? "").toString().split(";")[0].trim().toLowerCase();
  if (base === "") return "application/octet-stream";
  if (!/^[a-z0-9][a-z0-9.+-]*\/[a-z0-9][a-z0-9.+-]*$/.test(base)) return null;
  return ALLOWED_MIME.has(base) ? base : "application/octet-stream";
}

export function isPdf(mimeType: string | null | undefined, fileName?: string | null): boolean {
  if (normalizeMimeType(mimeType) === "application/pdf") return true;
  return /\.pdf$/i.test(fileName ?? "");
}

/** Parses the bridge's JSON reply; anything unparseable is a failure, never a silent success. */
export function resultFromJson(json: unknown): BridgeResult {
  if (typeof json !== "string") return { ok: false, error: "io_error" };
  try {
    const parsed = JSON.parse(json) as { ok?: unknown; error?: unknown; fileName?: unknown };
    if (parsed && parsed.ok === true) {
      return typeof parsed.fileName === "string" ? { ok: true, fileName: parsed.fileName } : { ok: true };
    }
    return { ok: false, error: typeof parsed?.error === "string" && parsed.error ? parsed.error : "io_error" };
  } catch {
    return { ok: false, error: "io_error" };
  }
}

export function userMessageForError(code: string): string {
  return USER_MESSAGES[code] ?? "The file couldn't be shared. Please try again.";
}

/** Chunked so large files don't overflow the argument limit of String.fromCharCode. */
export async function blobToBase64(blob: Blob): Promise<string> {
  const bytes = new Uint8Array(await blob.arrayBuffer());
  const CHUNK = 0x8000;
  let binary = "";
  for (let i = 0; i < bytes.length; i += CHUNK) {
    binary += String.fromCharCode.apply(null, Array.from(bytes.subarray(i, i + CHUNK)));
  }
  return btoa(binary);
}

async function callBridge(
  method: "saveAndShare" | "openWithViewer",
  blob: Blob,
  fileName: string,
  mimeType: string | undefined,
): Promise<void> {
  const bridge = typeof window !== "undefined" ? window.ElfAndroidFiles : undefined;
  if (!bridge || typeof bridge[method] !== "function") {
    throw new AndroidFileBridgeError("bridge_unavailable", userMessageForError("bridge_unavailable"));
  }
  if (blob.size === 0) throw new AndroidFileBridgeError("empty_data", userMessageForError("empty_data"));
  if (blob.size > MAX_BYTES) throw new AndroidFileBridgeError("too_large", userMessageForError("too_large"));

  const mime = normalizeMimeType(mimeType || blob.type);
  if (mime === null) throw new AndroidFileBridgeError("bad_mime", userMessageForError("bad_mime"));

  const base64 = await blobToBase64(blob);
  const result = resultFromJson(bridge[method](base64, sanitizeFileName(fileName, mime), mime));
  if (!result.ok) throw new AndroidFileBridgeError(result.error, userMessageForError(result.error));
}

/** Opens the Android share sheet (which also offers Save to Files/Drive). */
export function saveAndShareOnAndroid(blob: Blob, fileName: string, mimeType?: string): Promise<void> {
  return callBridge("saveAndShare", blob, fileName, mimeType);
}

/** Opens the file in an installed viewer (PDF reader, gallery, ...). */
export function openWithViewerOnAndroid(blob: Blob, fileName: string, mimeType?: string): Promise<void> {
  return callBridge("openWithViewer", blob, fileName, mimeType);
}

/** Message to show for any error thrown while sharing/opening on Android. */
export function androidErrorMessage(err: unknown): string {
  return err instanceof Error && err.message ? err.message : "The file couldn't be shared. Please try again.";
}

File: android/app/src/main/java/com/elitelevelfundraising/team/AndroidFileBridge.java
package com.elitelevelfundraising.team;

import android.content.ActivityNotFoundException;
import android.content.ClipData;
import android.content.Intent;
import android.net.Uri;
import android.util.Base64;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;
import android.widget.Toast;
import androidx.core.content.FileProvider;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.Callable;
import java.util.concurrent.FutureTask;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;
import org.json.JSONObject;

/**
 * Exposed to the page as window.ElfAndroidFiles. The Android WebView has no download manager,
 * no Web Share API and no PDF renderer, so the web layer hands file bytes (base64) here; they are
 * written to the app cache and handed to the system share sheet (saveAndShare) or to an installed
 * viewer (openWithViewer). Every method returns a JSON string {"ok":true|false,"error":"..."} and
 * never throws into JavaScript.
 *
 * Threading: @JavascriptInterface methods run on WebView's JavaBridge thread, so decoding and
 * file I/O happen there (off the UI thread). Anything touching the WebView, Toast or
 * startActivity is posted to the UI thread and awaited with a timeout.
 *
 * Exposure: addJavascriptInterface objects are visible to every frame in the page, so each call
 * first checks that the top-level page is https and on the configured server host. A third-party
 * frame embedded in an ELF page could still call it; the impact is limited to opening a share
 * sheet / viewer for a file the caller supplied itself (nothing is read from the device and
 * nothing is sent anywhere).
 */
public class AndroidFileBridge {

    static final String JS_NAME = "ElfAndroidFiles";
    private static final String CACHE_SUBDIR = "shared_files";
    private static final long MAX_BYTES = 25L * 1024 * 1024;
    private static final long STALE_MS = 24L * 60 * 60 * 1000;
    private static final int MAX_NAME_LENGTH = 100;
    private static final long UI_TIMEOUT_SECONDS = 5;
    private static final Pattern MIME_PATTERN = Pattern.compile("^[a-z0-9][a-z0-9.+-]*/[a-z0-9][a-z0-9.+-]*$");
    private static final Set<String> ALLOWED_MIME = new HashSet<>(Arrays.asList(
        "application/pdf",
        "text/csv",
        "text/calendar",
        "image/png",
        "image/jpeg",
        "image/webp",
        "image/gif",
        "application/octet-stream"
    ));

    private final MainActivity activity;

    AndroidFileBridge(MainActivity activity) {
        this.activity = activity;
    }

    @JavascriptInterface
    public String saveAndShare(String base64Data, String fileName, String mimeType) {
        return handle(base64Data, fileName, mimeType, false);
    }

    @JavascriptInterface
    public String openWithViewer(String base64Data, String fileName, String mimeType) {
        return handle(base64Data, fileName, mimeType, true);
    }

    private String handle(String base64Data, String fileName, String mimeType, boolean view) {
        try {
            if (!isTrustedOrigin()) return error("forbidden_origin");

            String mime = normalizeMimeType(mimeType);
            if (mime == null) return error("bad_mime");
            if (base64Data == null || base64Data.isEmpty()) return error("empty_data");
            if (base64Data.length() > (MAX_BYTES * 4 / 3) + 16) return error("too_large");

            byte[] bytes;
            try {
                bytes = Base64.decode(base64Data, Base64.DEFAULT);
            } catch (IllegalArgumentException e) {
                return error("bad_data");
            }
            if (bytes.length == 0) return error("empty_data");
            if (bytes.length > MAX_BYTES) return error("too_large");

            String safeName = sanitizeFileName(fileName, mime);
            File file;
            try {
                file = writeToCache(bytes, safeName);
            } catch (IOException | SecurityException e) {
                return error("io_error");
            }

            Uri uri = FileProvider.getUriForFile(activity, activity.getPackageName() + ".fileprovider", file);
            // A file with no usable type cannot be "viewed" (the platform accepts the intent but shows
            // nothing), so it goes to the share/save sheet instead of failing silently.
            boolean asView = view && !"application/octet-stream".equals(mime);
            return runOnUi(() -> launch(uri, safeName, mime, asView), error("launch_failed"));
        } catch (Throwable t) {
            return error("io_error");
        }
    }

    // Runs on the UI thread.
    private String launch(Uri uri, String name, String mime, boolean view) {
        try {
            if (view) {
                Intent open = new Intent(Intent.ACTION_VIEW);
                open.setDataAndType(uri, mime);
                open.setClipData(ClipData.newRawUri(name, uri));
                open.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                try {
                    activity.startActivity(open);
                } catch (ActivityNotFoundException e) {
                    Toast.makeText(
                        activity,
                        "No app installed to open this file. Try saving it instead.",
                        Toast.LENGTH_LONG
                    ).show();
                    return error("no_handler");
                }
            } else {
                Intent send = new Intent(Intent.ACTION_SEND);
                send.setType(mime);
                send.putExtra(Intent.EXTRA_STREAM, uri);
                send.setClipData(ClipData.newRawUri(name, uri));
                send.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                activity.startActivity(Intent.createChooser(send, "Share or save file"));
            }
            JSONObject ok = new JSONObject();
            ok.put("ok", true);
            ok.put("fileName", name);
            return ok.toString();
        } catch (Throwable t) {
            return error("launch_failed");
        }
    }

    /** True only when the top-level page is https on the configured server host. */
    private boolean isTrustedOrigin() {
        return runOnUi(() -> {
            if (activity.getBridge() == null) return false;
            WebView webView = activity.getBridge().getWebView();
            String serverUrl = activity.getBridge().getServerUrl();
            if (webView == null || webView.getUrl() == null || serverUrl == null) return false;
            Uri page = Uri.parse(webView.getUrl());
            Uri server = Uri.parse(serverUrl);
            return "https".equalsIgnoreCase(page.getScheme())
                && page.getHost() != null
                && page.getHost().equalsIgnoreCase(server.getHost());
        }, false);
    }

    private <T> T runOnUi(Callable<T> task, T fallback) {
        FutureTask<T> future = new FutureTask<>(task);
        activity.runOnUiThread(future);
        try {
            return future.get(UI_TIMEOUT_SECONDS, TimeUnit.SECONDS);
        } catch (Exception e) {
            return fallback;
        }
    }

    /** Writes into cache/shared_files/<unique>/<name>, after purging stale entries. */
    private File writeToCache(byte[] bytes, String safeName) throws IOException {
        File root = new File(activity.getCacheDir(), CACHE_SUBDIR);
        if (!root.exists() && !root.mkdirs()) throw new IOException("cache dir");
        purgeStale(root);

        File dir = new File(root, System.currentTimeMillis() + "-" + UUID.randomUUID().toString().substring(0, 8));
        if (!dir.mkdirs()) throw new IOException("cache subdir");
        File file = new File(dir, safeName);
        // The name is already sanitized; this is a second guard against escaping the cache dir.
        if (!file.getCanonicalPath().startsWith(root.getCanonicalPath() + File.separator)) {
            throw new IOException("path escape");
        }
        try (FileOutputStream out = new FileOutputStream(file)) {
            out.write(bytes);
        }
        return file;
    }

    private void purgeStale(File root) {
        File[] children = root.listFiles();
        if (children == null) return;
        long cutoff = System.currentTimeMillis() - STALE_MS;
        for (File child : children) {
            if (child.lastModified() < cutoff) deleteRecursively(child);
        }
    }

    private void deleteRecursively(File file) {
        File[] children = file.listFiles();
        if (children != null) {
            for (File child : children) deleteRecursively(child);
        }
        //noinspection ResultOfMethodCallIgnored
        file.delete();
    }

    static String sanitizeFileName(String raw, String mime) {
        String name = raw == null ? "" : raw;
        int slash = Math.max(name.lastIndexOf('/'), name.lastIndexOf('\\'));
        if (slash >= 0) name = name.substring(slash + 1);
        name = name.replaceAll("[^A-Za-z0-9._ -]", "_");
        name = name.replaceAll("\\.{2,}", ".");
        name = name.replaceAll("^[. ]+", "").trim();
        if (name.length() > MAX_NAME_LENGTH) {
            int dot = name.lastIndexOf('.');
            String ext = (dot > 0 && name.length() - dot <= 10) ? name.substring(dot) : "";
            name = name.substring(0, MAX_NAME_LENGTH - ext.length()) + ext;
        }
        if (name.isEmpty()) name = "file";
        if (!name.contains(".")) {
            String ext = extensionFor(mime);
            if (ext != null) name = name + "." + ext;
        }
        return name;
    }

    /** null = malformed; well-formed but unlisted types become application/octet-stream. */
    static String normalizeMimeType(String raw) {
        String base = raw == null ? "" : raw.split(";", 2)[0].trim().toLowerCase(Locale.US);
        if (base.isEmpty()) return "application/octet-stream";
        if (!MIME_PATTERN.matcher(base).matches()) return null;
        return ALLOWED_MIME.contains(base) ? base : "application/octet-stream";
    }

    private static String extensionFor(String mime) {
        switch (mime) {
            case "application/pdf": return "pdf";
            case "text/csv": return "csv";
            case "text/calendar": return "ics";
            case "image/png": return "png";
            case "image/jpeg": return "jpg";
            case "image/webp": return "webp";
            case "image/gif": return "gif";
            default: return null;
        }
    }

    private static String error(String code) {
        try {
            JSONObject json = new JSONObject();
            json.put("ok", false);
            json.put("error", code);
            return json.toString();
        } catch (Exception e) {
            return "{\"ok\":false,\"error\":\"" + code + "\"}";
        }
    }
}
